<launch>
	<!-- FROBIT PARAMETERS --> 
	<param name="diff_steer_wheel_distance" value="0.375"/>
	<param name="ticks_per_meter_left" value="677" />
	<param name="ticks_per_meter_right" value="677" />

	<!-- FROBIT INTERFACE -->
	<group ns="fmControllers">
		<!--node pkg="sdu_frobit_v2" name="frobit" type="frobit_interface_node.py" output="screen" -->
		<node pkg="mr_frobit_interface" name="frobit" type="frobit_interface_node.py" output="screen" >
			<param name="actuation_enable_sub" value="/fmSafe/actuation_enable" />
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="/fmCommand/cmd_vel" value="True" />
			<param name="enc_left_pub" value="/fmInformation/encoder_left" />
			<param name="enc_right_pub" value="/fmInformation/encoder_right" />
			<param name="wheel_status_left_pub" value="/fmInformation/wheel_status_left" />
			<param name="wheel_status_right_pub" value="/fmInformation/wheel_status_right" />
			<param name="wheel_feedback_left_pub" value="/fmInformation/wheel_feedback_left" />
			<param name="wheel_feedback_right_pub" value="/fmInformation/wheel_feedback_right" />
			<param name="wheel_pid_left_pub" value="/fmInformation/wheel_pid_left" />
			<param name="wheel_pid_right_pub" value="/fmInformation/wheel_pid_right" />
			<param name="nmea_from_frobit_sub" value="/fmSignal/nmea_from_frobit" />	 
			<param name="nmea_to_frobit_pub" value="/fmSignal/nmea_to_frobit" />	 
			<param name="publish_wheel_status_rate" value="10" />
			<param name="publish_wheel_feedback_rate" value="10" />
			<param name="publish_wheel_pid_rate" value="10" />
			<param name="castor_front" value="true"/>
		    <param name="max_linear_acceleration" value="16.2"/>
        	<param name="max_angular_acceleration" value="16.0"/>
			<param name="wheel_drive_feed_forward" value="85" /> 
			<param name="wheel_drive_kp" value="5" /> 
			<param name="wheel_drive_ki" value="10" /> 
			<param name="wheel_drive_kd" value="5" /> 
			<param name="wheel_drive_max_integral_output" value="300" /> 
			<param name="wheel_turn_feed_forward" value="110" /> 
			<param name="wheel_turn_kp" value="10" /> 
			<param name="wheel_turn_ki" value="20" /> 
			<param name="wheel_turn_kd" value="3" /> 
			<param name="wheel_turn_max_integral_output" value="300" /> 
			<param name="supply_voltage_scale_factor" value="0.03747" /> 
			<param name="min_supply_voltage" value="11.7" /> 
			<param name="show_voltage_interval" value="5" /> 
		</node>
	</group>

	<!-- SAFETY MODULE -->
	<group ns="fmSafety">
		<node pkg="mr_incident_handler" name="incident_handler" type="incidentHandlerNode.py" output="screen">
    		<param name="deadman_enable" value="True" />
    		<param name="deadman_timeout" value="0.100" />
    		<param name="deadman_sub" value="/fmSafe/deadman" />
    		<param name="critical_fault_enable" value="True" />
    		<param name="critical_fault_timeout" value="0.100" />
    		<param name="critical_fault_sub" value="/fmSafe/critical_fault" />
	    	<param name="obstacle_enable" value="True" />
	    	<param name="obstacle_sub" value="/mrObstacleDetector/status" />
			<param name="actuation_enable_pub" value="/fmSafe/actuation_enable"/>
			<param name="nav_currentnode_sub" value="/mrNavigationController/currentNode"/>
		</node>
	</group>

	<!-- FROBIT SERIAL DEVICE AND PROTOCOL -->
	<node pkg="serial_nmea" name="frobit_nmea" type="serial_nmea_node" respawn="true" output="screen">
		<param name="nmea_from_device_pub" value="/fmSignal/nmea_from_frobit"/>
		<param name="nmea_to_device_sub" value="/fmSignal/nmea_to_frobit"/>
		<param name="serial_device" value="/dev/frobit"/>
		<param name="serial_baudrate" value="57600"/>
		<param name="use_nmea_checksum" value="true"/>
        </node>

	<!-- TWIST CONVERTER -->
<!--
	Dominik: i think we dont use this one
	<group ns="fmLib">
		<node pkg="twist_stamped_to_twist" name="converter" type="cmd_vel_converter.py">	
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="cmd_vel_pub" value="/cmd_vel" />
		</node>
	</group>
 
-->
	<node pkg="twist_to_twist_stamped" name="converter_stamped" type="cmd_vel_converter.py">	
		<param name="cmd_vel_sub" value="/cmd_vel" />
		<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
	</node>

	<!-- CALCULATE ODOMETRY (Based on wheel encoders and IMU yaw axis orientation) -->
	<group ns="fmProcessors">
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/encoder_left" />
			<param name="enc_right_sub" value="/fmInformation/encoder_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_reset_sub" value="/fmInformation/odom_reset" />
			<param name="odom_pub" value="/fmKnowledge/pose" />
			<param name="odom_estimate_frame" value="odom" />
			<param name="vehicle_frame" value="base_link" />
			<param name="encoder_output" value="absolute" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="z" />
		</node>
	</group>

	<!-- IMU SERIAL COMMUNICATION -->
	<node pkg="serial_string" name="imu_serial" type="serial_string_node" respawn="true" output="screen">
		<param name="publisher_topic" value="/fmData/imu_rx" />	 
		<param name="subscriber_topic" value="/fmData/imu_tx" />
		<param name="device" value="/dev/imu" />	
		<param name="baudrate" value="115200" />
		<param name="termination_character" value="10"/>
	</node> 

	<!-- Extract data from messages transmitted by the VectorNav VN-100 IMU -->
	<node pkg="vectornav_vn100" name="imu_node"  type="vectornav_vn100_node" output="screen"  respawn="true">
		<param name="vectornav_vn100_sub" value="/fmData/imu_rx"/>
		<param name="imu_pub" value="/fmInformation/imu"/>
		<param name="frame_id" value="imu_link"/>
		<param name="use_enu" value="true"/>
		<param name="covariance_x" value="0.001"/>
		<param name="covariance_y" value="0.001"/>
		<param name="covariance_z" value="0.001"/>
	 </node>

	<!-- LIDAR -->
	<node name="sick_tim310s01" pkg="sick_tim" type="sick_tim310s01" respawn="false" output="screen">
	    <param name="min_ang" type="double" value="-1.6" />
	    <param name="max_ang" type="double" value="1.6" />
	    <param name="intensity" type="bool" value="True" />
	    <param name="skip" type="int" value="0" />
	    <param name="frame_id" type="str" value="base_laser_link" />
	    <param name="time_offset" type="double" value="-0.001" />
	    <param name="publish_datagram" type="bool" value="False" />
	    <param name="port" type="str" value="/dev/lidar" />
	</node>

	<!-- MR OBSTACLE DETECTOR -->
	<group ns="MR_obstacle_detector">
		<node pkg="mr_obstacle_detector" name="mr_obstacle_detector" type="obstacle_detector.py">
			<param name="threshold_proximityAlert" value="0.4" />
			<param name="threshold_colliding" value="0.3" />
			<param name="threshold_ignore" value="0.05" />
			<param name="laser_scan" value="/scan" />
			<param name="publishTopic" value="/mrObstacleDetector/status" />
			<param name="enabler" value="/mrObstacleDetector/enabler" />
		</node>
	</group>

	<!-- MR TIP CONTROLLER -->
	<node pkg="mr_tip_controller" name="mr_tip_controller" type="mr_tip_controller" output="screen">
		<param name="baud_rate" value="115200"/>
		<param name="port" value="/dev/serial/by-id/usb-Arduino_Srl_Arduino_Uno_7543932393535120F172-if00"/>  	
		<param name="debug" value="true"/>  
		<param name="tipper_srv" value="/mrTipController/tip"/>	   	
	</node>

	<!-- MR HMI -->
	<group ns="MR_HMI">
		<node pkg="mr_hmi" name="HMI" type="hmiNode.py" output="screen">
			<param name="mr_hmi_status_sub" value="/mrHMI/status"/>
			<param name="mr_hmi_run_pub" value="/mrHMI/run"/>
			<param name="deadman_pub" value="/fmSafe/deadman"/>
			<param name="cmd_pub" value="/fmCommand/cmd_vel"/>
			<param name="tipper_srv" value="/mrTipController/tip"/>
		</node>			
	</group>

	<!-- MR Button -->
	<group ns="MR_Button">
		<node pkg="mr_button" name="Button" type="mr_button" output="screen">
			<param name="baud_rate" value="115200"/>
			<param name="port" value="/dev/serial/by-id/usb-Texas_Instruments_In-Circuit_Debug_Interface_0E203B83-if00"/> 
			<param name="mr_collision_sub" value="/mrObstacleDetector/status"/>
			<param name="mr_button_pub" value="/mrButton/run"/> 
			<param name="mr_button_sub" value="/mrButton/status"/> 	
			<param name="mr_main_mode_sub" value="/mrMain/mode"/>
		</node>
	</group>

	<!-- MR Main -->
	<group ns="MR_Main">
		<node pkg="mr_main" name="Main" type="mr_main" output="screen">
			<param name="nav_perform_srv" value="MR_NavigationController/performAction"/> 
			<param name="nav_status_sub" value="mrNavigationController/status"/> 
			<param name="nav_currentnode_sub" value="mrNavigationController/currentNode"/>	
			<param name="button_sub" value="/mrButton/run"/> 
			<param name="hmi_sub" value="mrHMI/run"/> 
			<param name="hmi_pub" value="mrHMI/status"/> 
			<param name="tipper_srv" value="mrTipController/tip"/> 
			<param name="mes_pub" value="mrMESClient/msgToServer"/> 
			<param name="mes_sub" value="mrMESClient/msgFromServer"/> 
			<param name="obstacle_detector_sub" value="/mrObstacleDetector/status"/> 
			<param name="battery_sub" value="/fmInformation/battery"/>
			<param name="mode_pub" value="/mrMain/mode"/>
			<param name="check_battery" value="false"/>
			<param name="battery_low" value="12.4"/> 
			<param name="battery_critic" value="12.0"/> 
			<param name="desired_charge" value="14.1"/> 
		</node>
	</group>

	<!-- MR Go -->
	<group ns="MR_Go">
    		<node pkg="mr_go" name="mr_go" type="mr_go" respawn="true" output="screen">
		    	<param name="odometry" value="/fmKnowledge/pose"/>
		    	<param name="pub_twist" value="/fmCommand/cmd_vel"/>
		    	<param name="pub_deadman" value="/fmSafe/deadman"/>
		    	<param name="srv_move" value="/mrGo/move"/>
		    	<param name="linear_precision" value="0.01"/>
		    	<param name="angular_precision" value="0.500"/>
		    	<param name="linear_speed" value="0.2"/>
		    	<param name="angular_speed" value="0.3"/>
    		 </node>
	</group>

	<!-- MR Camera -->
	<group ns="MR_Camera">
   		<node pkg="mr_camera" name="camera" type="mr_camera" output="screen">
    		</node>
	</group>

	<!-- MR Camera Processing -->
	<group ns="MR_CameraProcessing">
	    	<node pkg="mr_camera_processing" name="CameraProcessing" type="mr_camera_processing" output="screen">
			<param name="sub_image" value="/mrCamera/image" />
			<param name="pub_cross" value="/mrCameraProcessing/cross" />
			<param name="pub_qr" value="/mrCameraProcessing/QR" />
			<param name="pub_line" value="/mrCameraProcessing/line" />
			<param name="pub_image" value="/mrCameraProcessing/output_image" />
			<param name="srv_enable" value="/mrCameraProcessing/enable" />
		</node>
	</group>

	<!-- MR Line Follower -->
	<group ns="MR_LineFollower">
		<node pkg="mr_line_follower" name="LineFollower" type="mr_line_follower" output="screen">
			<param name="pid_p" value="0.8" />
			<param name="pud_i" value="0.5" />
			<param name="pid_d" value="0.5" />
			<param name="pid_dt" value="0.33" />
			<param name="pid_max" value="320" />
			<param name="pid_min" value="-320" />
			<param name="reference_point_x" value="320" />
			<param name="reference_point_y" value="240" />
			<param name="robot_speed" value="0.14" />
			<param name="lidar_distance" value="0.2" />
			<param name="relative_distance" value="0.2" />
		
			<param name="srv_mr_camera_processing_enable_name" value="/mrCameraProcessing/enable" />
			<param name="sub_cross" value="/mrCameraProcessing/cross" />
			<param name="sub_line" value="/mrCameraProcessing/line" />
			<param name="sub_lidar" value="/scan" />
			<param name="sub_odom" value="/fmKnowledge/pose" />
			<param name="sub_qr" value="/mrCameraProcessing/QR" />
			<param name="pub_twist" value="/fmCommand/cmd_vel" />
			<param name="pub_deadman" value="/fmSafe/deadman" />
		</node>
	</group>


	<!-- MR MES Client -->
	<group ns="MR_MES_Client">
		<node pkg="mr_mes_client" name="mr_mes_client" type="mr_mes_client">
			<param name="mesPub" value="/mrMESClient/msgFromServer"/>
			<param name="mesSub" value="/mrMESClient/msgToServer"/>
			<param name="serverIP" value="10.115.253.233"/>
			<param name="serverPort" value="21240"/>
		</node>
	</group>	

	<!-- MR Navigation Controller -->
	<node pkg="mr_navigation_controller" name="mrNavigationController" type="mr_navigation_controller" output="screen">
		<param name="lineFollowEnableService" value="mrLineFollower/lineUntilQR"/>
		<param name="lineFollowEnableLidarService" value="mrLineFollower/lineUntilLidar"/>
		<param name="lineFollowEnableRelativeService" value="mrLineFollower/lineUntilRelative"/>
		<param name="obstacleDetectorService" value="/mrObstacleDetector/enabler"/>
		<param name="moveService" value="mrGo/move"/>
		<param name="performAction" value="mrNavigationController/performAction"/>
		<param name="status" value="mrNavigationController/status"/>
		<param name="currentNode" value="mrNavigationController/currentNode"/>
		<param name="searchLimit" value="100"/>
		<param name="start_node" value="line_start"/>
	</node>

	<node pkg="tf" type="static_transform_publisher" name="laser_transform" args="0.25 0 0 0 0 0 base_link base_laser_link 50"  >
	</node>

	<node pkg="tf" type="static_transform_publisher" name="imu_transform" args="0.10 0 0 0 0 0 base_link imu_link 50"  >
	</node>

	<node pkg="map_server" type="map_server" name="map_server" args="/home/frobit_pro_group1/roswork/src/frobo_rsd//Navigation_Setup/maps/frontal_map_big_edited.yaml" output="screen"/>


<!--
	<node pkg="map_server" type="map_server" name="map_server" args="/home/frobit_pro_group1/roswork/src/frobo_rsd//Navigation_Setup/maps/entire_map.yaml" output="screen"/>
-->


	<node pkg="amcl" type="amcl" name="amcl" output="screen">
		<param name="odom_alpha1" value="0.8"/>
		<param name="odom_alpha2" value="0.8"/>
		<param name="odom_alpha3" value="0.8"/>
		<param name="odom_alpha4" value="0.8"/>
		<param name="min_particles" value="30"/>

		<param name="initial_pose_a" value="0"/>
		<param name="initial_pose_x" value="0"/>
		<param name="initial_pose_y" value="0"/>
	
		<param name="base_frame_id" value="base_link"/>
		<param name="odom_frame_id" value="odom"/>

	</node>


	<node pkg="move_base" name="move_base" type="move_base" output="screen">
		<rosparam file="/home/frobit_pro_group1/roswork/src/frobo_rsd/Navigation_Setup/move_base_config/move_base_params.yaml" command="load" />
    		<param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS " />  
		<rosparam file="/home/frobit_pro_group1/roswork/src/frobo_rsd/Navigation_Setup/move_base_config/costmap_common_params.yaml" command="load" ns="global_costmap"/>
		<rosparam file="/home/frobit_pro_group1/roswork/src/frobo_rsd/Navigation_Setup/move_base_config/costmap_common_params.yaml" command="load" ns="local_costmap"/>
		<rosparam file="/home/frobit_pro_group1/roswork/src/frobo_rsd/Navigation_Setup/move_base_config/local_costmap_params.yaml" command="load"/>
		<rosparam file="/home/frobit_pro_group1/roswork/src/frobo_rsd/Navigation_Setup/move_base_config/global_costmap_params.yaml" command="load"/>
		<rosparam file="/home/frobit_pro_group1/roswork/src/frobo_rsd/Navigation_Setup/move_base_config/dwa_local_planner_params.yaml" command="load"/>
	</node>


</launch>

